### ISSUE TITLE
Cylindrical face selection

### ISSUE THREAD

**Issue by bradymadams** [tetonsim/is-cura-ui#133]
Detect holes and cylindrical cross sections in the face selection logic.

**Comment by jake-kirkland-teton** [comment]
So I'm not exactly sure what the desired behavior is in all cases, but in this STL it looks like the cylinder selection works well in some places, but not in some others.

The inside of this cylinder selects correctly, but not the outside? Is that the correct behavior here?

[CylinderMazeLid.zip](https://github.com/tetonsim/is-cura-ui/files/4645652/CylinderMazeLid.zip)

**Comment by dougkenik** [comment]
It works "mostly" well/ Here are some screenshots and a file where it  breaks down. 
![image](https://user-images.githubusercontent.com/24813209/82243114-dff3e780-98f3-11ea-8d69-079faccad4c3.png)
![image](https://user-images.githubusercontent.com/24813209/82243173-fc901f80-98f3-11ea-9df0-2081452a0a30.png)
![image](https://user-images.githubusercontent.com/24813209/82243212-0fa2ef80-98f4-11ea-9c95-33852a7f53f1.png)

[Laser Base.zip](https://github.com/tetonsim/is-cura-ui/files/4645670/Laser.Base.zip)

**Comment by rickdalgarno** [comment]
Cura crashes when I try to select the face shown below. This is the "Gimbal Base.stl". The error messages from Cura and the stderr.log files are attached.

[stderr.log](https://github.com/tetonsim/is-cura-ui/files/4646327/stderr.log)
[error_log.txt](https://github.com/tetonsim/is-cura-ui/files/4646329/error_log.txt)
![gimbal_base](https://user-images.githubusercontent.com/34459028/82252931-37528180-990d-11ea-8fa7-8f1159778275.png)

**Comment by rickdalgarno** [comment]
I created a new STL that has holes with different shapes and sizes. Some of the holes are on a planar region and the other holes are on a curved section. 

Some observations for selecting holes in the planar section:

- The small, medium, and large holes are work as expected
- The cone shaped hole does not work
- The ellipse will grab up to about 40% of the surface. Depends on where you click.
- The hole with a fillet does not work but the hole with the chamfer does work.

Some observation for selecting holes in the curved section:

- The 3 holes sometimes work and sometimes don't. It just depends on where you click in the hole.
- The ellipse behavior is similar to the planar ellipse.

In general, the faces that end up being selected depend on where you click in the hole. 

![hole tester](https://user-images.githubusercontent.com/34459028/82257917-1d696c80-9916-11ea-9f4e-eda60d387445.png)

**Comment by thopiekar** [comment]
@rickdalgarno Could you send me the "Gimbal Base.stl"?

Generally, many thanks for testing. The cylinder detection works on cylinders only. Cones and ellipses won't work since the algorithm, which has been originally developed by @bradymadams, guesses a full cylinder using the clicked triangle and a neighbored triangle on the mesh. Using this cylinder a "tolerance zone" around the round shape is being created and all triangles, which are fully matched within this cylindrical zone, will be selected. I guess it is quite self explaining what will happen, if the same mechanism is executed on a cone or ellipsis. :)

Based on the given data, I'll review the meshes and see how I can tweak the selection :)

**Comment by rickdalgarno** [comment]
@thopiekar Gimbal Base.stl attached
[Gimbal Base.zip](https://github.com/tetonsim/is-cura-ui/files/4684441/Gimbal.Base.zip)

**Comment by thopiekar** [comment]
Thank you, @rickdalgarno !

@dougkenik: Regarding your Laser Base: Every second of the holes on the bottom looks really messy. I can imagine that our algorithms detects more than two triangles which would fit onto a plane and believes it is a plane instead of a cylinder. I haven't seen such a fragmented cylinder mesh before.
What I mean as fragmented:
![image](https://user-images.githubusercontent.com/1847437/82940018-7c715600-9f94-11ea-86ac-3477377e868e.png)
Versus a mesh that should be nicely detected as cylinder:
![image](https://user-images.githubusercontent.com/1847437/82940081-990d8e00-9f94-11ea-8c60-65aa74912cbb.png)

**Comment by dougkenik** [comment]
Yeah, this is the reality of what we're working with though. We should try to catch these types of things, if at all possible. If not, we should investigate other methods

**Comment by thopiekar** [comment]
Hmm.. I might need to brainstorm with @bradymadams how to proceed here. Whether we really want to refine our current implementation or whether parts need to be rethought. :thinking:

**Comment by bradymadams** [comment]
@thopiekar we'll need to think if there's a way we can identify that situation. I don't have any ideas off the top of my head, but I will give it some thought.

Overall, I think we should aggregate the "failures" from the testing that was done and then work with Doug to determine which cylinders need to be detected that currently are not, in order to deliver a satisfactory cylinder detection experience. I'll work on summarizing the results from everyone's testing and then we can go from there.

**Comment by bradymadams** [comment]
After review I think we can put the failures into 4 categories:

1. The outside of the maze cylinder has a very inconsistent triangulation. I would vote to put this at low priority.

![image](https://user-images.githubusercontent.com/23110600/83092529-bad14880-a05a-11ea-9143-5d09cef9ccc7.png)

2. The gimbal base crash should be looked into and fixed, assuming we move forward with this approach.

3. The hole with the fillet is triangulated nicely. I think the triangle neighbor search is picking up the triangles in the fillet and throwing it off.

![image](https://user-images.githubusercontent.com/23110600/83093021-e4d73a80-a05b-11ea-86aa-dc90da933b1f.png)

4. The rest have slightly inconsistent triangulations as @thopiekar pointed out. These, in my opinion, will be the make or break for this approach. If we want to move forward with this approach we should probably look into these first to see if we can detect these cylinders. If it is not feasible, then we may need to explore other options.

**Comment by dougkenik** [comment]
Let's dive into the fourth point to figure out if we can find a way to identify these cylinders.

**Comment by thopiekar** [comment]
(2) Fixed. It was a numerical issue. The value, which is passed to math.asin was outside [-1.0, 1.0]. To be exact the value was `1.0000000000000002`. I fixed the issue in that way, that I generally assume that everything calculated by this function is built on a trustable equation.
Therefore, every value that is outside the boundaries for math.asin, will be corrected to the lowest or highest value of the boundary. Tested it with @rickdalgarno gimbal base and it works fine for all the 4 holes:
![image](https://user-images.githubusercontent.com/1847437/83421425-61856280-a428-11ea-8f83-45c5dce12dd9.png)

(3) I'm continuing there now. Keeping you posted.

**Comment by thopiekar** [comment]
Done. Technical information is included in the commit messages.
These three changes are fixing (2), (3) and the issue mentioned by @dougkenik.

Does anyone of you see other problems which need to be fixed within this task? :wink: 

See:
![image](https://user-images.githubusercontent.com/1847437/83424267-79f77c00-a42c-11ea-9e62-2a3c9860f290.png)
![image](https://user-images.githubusercontent.com/1847437/83429210-fe99c880-a433-11ea-9f9d-f431e8ad3003.png)

**Comment by bradymadams** [comment]
https://teton-cdn.s3-us-west-2.amazonaws.com/smartslice-cura-plugin/dist/SmartSlicePlugin-r20.1-cyl-select.15.curapackage

New build with fixes for items 2 and 3 from my last comment:

> 2. The gimbal base crash should be looked into and fixed, assuming we move forward with this approach.
> 3. The hole with the fillet is triangulated nicely. I think the triangle neighbor search is picking up the triangles in the fillet and throwing it off.

**Comment by thopiekar** [comment]
Has any further testing been done? Any remaining steps and tasks required to finish up the work here? 😁

**Comment by bradymadams** [comment]
@thopiekar I quickly tested your latest changes on the STLs that others had posted to this issue, but I'm not sure if anyone else has tested. However, I do know, we need to see if we can take a run at detecting cylinders on less regularly triangulated cylinders (item 4 on a previous comment of mine):

>The rest have irregular triangulations as @thopiekar pointed out. These, in my opinion, will be the make or break for this approach. If we want to move forward with this approach we should probably look into these first to see if we can detect these cylinders. If it is not feasible, then we may need to explore other options.

**Comment by thopiekar** [comment]
So we are looking for an alternative to the simple check, right? The one that says: "If a plane could be fitted using two planes only, then it is a cylinder."
I'm already brainstorming of an replacement. I'll share my ideas when I have two options or more at least.

**Comment by dougkenik** [comment]
Some feedback from the new version:

1. I received a fatal crash with the Gimbal_Base (screenshots below)
![image](https://user-images.githubusercontent.com/24813209/84155236-f5ed5780-aa1c-11ea-86b9-0480b276b88a.png)

2. On the Laser Base, we are still missing triangles:
![image](https://user-images.githubusercontent.com/24813209/84155545-49f83c00-aa1d-11ea-9edc-e6b21291426c.png)
![image](https://user-images.githubusercontent.com/24813209/84155612-6005fc80-aa1d-11ea-8de2-29e1935b3e87.png)

**Comment by bradymadams** [comment]
@dougkenik that traceback is showing the old pywim code. I think I created the last build wrong. Let me look into it.

**Comment by bradymadams** [comment]
Build 15 was built incorrectly. I think I accidentally chose the r20.0 cylinder pywim branch. Build 17 has the correct pywim branch. Sorry about that.

https://teton-cdn.s3-us-west-2.amazonaws.com/smartslice-cura-plugin/dist/SmartSlicePlugin-r20.1-cyl-select.17.curapackage

**Comment by dougkenik** [comment]
Ok - disregard my previous comments. New comments below. In general, it seems to be working much better. 

1. It still hangs up on surfaces which are cylindrical, but not perfect. In the below (laser_base), the surface is cylindrical, but the "z" edge isn't flat(it's wavy). IMO, we should be able to capture this
![image](https://user-images.githubusercontent.com/24813209/84157237-5aa9b180-aa1f-11ea-985b-21b275a5428c.png)

2. I know this surface is an ellipse, but I think we should be able to capture some of it, and we should definitely not be capturing the flat surface adjacent (laser_base)
![image](https://user-images.githubusercontent.com/24813209/84157602-c724b080-aa1f-11ea-8394-d8d896977eb5.png)

3. Again, the z-surface of this cylinder is not constant, but I think we should be able to select the cylinder:
![image](https://user-images.githubusercontent.com/24813209/84157737-efacaa80-aa1f-11ea-9ce1-8e8e76ab9e4e.png)

**Comment by rickdalgarno** [comment]
1. Working nicely on this hole. It is grabbing the semi-circle, which is what I think the expected behavior should be.

![image](https://user-images.githubusercontent.com/34459028/84177129-aec48e00-aa3f-11ea-88eb-0c1e7ae64e91.png)

2. The hole with the fillet from "hole tester v1.stl" now works as expected.

![image](https://user-images.githubusercontent.com/34459028/84177491-2c889980-aa40-11ea-992a-87748eda430d.png)

3. IKEA_hinge_jig_1.stl has some issues. The holes shown below are not selectable.

![image](https://user-images.githubusercontent.com/34459028/84178240-51314100-aa41-11ea-9f18-a657d170f671.png)

Here are the stl faces for the bad holes

![image](https://user-images.githubusercontent.com/34459028/84178390-95244600-aa41-11ea-810a-6caee47d7572.png)

4. LowerJig.stl has some issues as well. Most of the holes do not get fully selected and the amount of faces that do get selected depends on where you click in the hole. These are cylinders and should be selectable. Most of the stl faces look like this:

![image](https://user-images.githubusercontent.com/34459028/84178870-5c38a100-aa42-11ea-981e-f682603719f3.png)

[stl files.zip](https://github.com/tetonsim/is-cura-ui/files/4753646/stl.files.zip)

**Comment by rickdalgarno** [comment]
One more... the 4 small holes on this part are not selectable.

![image](https://user-images.githubusercontent.com/34459028/84179831-c43bb700-aa43-11ea-93fb-3647cdd9538a.png)

[Hole_Template.zip](https://github.com/tetonsim/is-cura-ui/files/4753709/Hole_Template.zip)

**Comment by thopiekar** [comment]
Ok, based on the latest feedback I made adjustments, so that the selection of the "DialIndicatorMountV3" works. After doing the adjustment there, "IKEA_hinge_jig_1" works now, too.
Looks like the boundary condition stated here was too tight - I added some additional lines to make the check more clear:
https://github.com/tetonsim/pywim/pull/38/commits/2b06fea9fc778d12a55140ce5639090290414f1a

Regarding the other issues, I don't have an ideal solution in mind. I did some analysis on the meshes, which I'd like to share with @bradymadams first. However, if you are interested I can share the info with you, too. It can be either on Slack or publically in this issue ticket here.
Basically, a solution needs to be found, which makes our algorithm less susceptible towards coarse tesseleration.
![Screenshot_20200614_170424](https://user-images.githubusercontent.com/1847437/84598835-c9e33380-ae6d-11ea-8f40-2a7f3e678172.png)

**Comment by bradymadams** [comment]
New build: https://teton-cdn.s3-us-west-2.amazonaws.com/smartslice-cura-plugin/dist/SmartSlicePlugin-r20.1-cyl-select.21.curapackage

**Comment by faisal-bhuiyan** [comment]
Played around a little bit with the new build of the SS plugin with cylinder selection package, it is working really nicely for several closed cylindrical surfaces I tried to select. The following two examples are I think might not be expected behavior from an user experience perspective:

a. The individual flat surfaces on the following part could not be selected.

![image](https://user-images.githubusercontent.com/59463700/86041924-abe80800-ba03-11ea-825a-00c94dfef6dd.png)


b. Also the whole convex surface is not being selected as an user might intend to as shown in the following image.
![image](https://user-images.githubusercontent.com/59463700/86042177-02554680-ba04-11ea-9fb1-e5d2af7cb47f.png)

**Comment by thopiekar** [comment]
Hello @faisal-bhuiyan,
I explained the reason in a private conversation with @bradymadams and @dougkenik. The reason is the following:
`Since cylinder are simplified as squares and then split into triangles, we also need to grow planar, too... Just to jump over to the other half of your square.` I'm going to look for a method to limit the growth in planar direction.

Regarding the other point, there is function which does a growth in convex, too, but at the moment there is no way to select the selection mode. At least it is not clear (at least to me) where it will be a combox to select from or an automatic solution (which I personally find risky unless we are not sure the auto-mode is 100% reliable in every situation).
